### PR TITLE
Param find macro

### PR DESCRIPTION
**Pull request by stephanbro**
This pull request adds a macro for looking up parameter handles instead of doing a string lookup. The main advantage of this is that parameters that there is now the ability to have a compile time check for if a parameter exists in the build. This is now possible since the parameter order is known at compile time as of #6254.

The number of `param_find()` function calls will now go down to items requested externally by string (i.e. over mavlink), strings generated in loops (like when using scanf), and when needing to handle the possibility of a parameter not existing. Additionally, there should also be a boot time speed up, as much less `param_find()` functions are called.

List of changes:
* Generate the parameter handles at compile time
* Add a `PARAM_FIND` macro for getting the handle directly
* Changed most `param_find()` function calls to `PARAM_FIND()` macro calls
* Moved ATT_VIBE_THRESH and SDLOG_UTC_OFFSET parameters to more appropriate modules
* msg_gen now depends on param_gen as a result of systemlib/param/param.h now including the generated px4_parameters.h

Keeping track of which parameters are used will now only work with string lookups. Does this still make sense to keep that? Could a script accomplish something similar?

I've only built the px4fmu_v4-default, px4fmu_v2-default, and the SITL builds so far. More tweaks will probably be required with the other builds.

### CODE REVIEW

**Comment by dagar** [comment]
If you want to build all the nuttx configurations locally try `make qgc_firmware` and `make alt_firmware` or look at the semaphoreci failures. https://semaphoreci.com/px4/firmware/branches/pull-request-6283/builds/1

**Comment by LorenzMeier** [comment]
The goal of param find marking used params was to avoid Instantiating params that are not used in the current configuration. You can check if you broke that by loading the params in QGC and checking that you are not getting fixed wing params on a multicopter.

**Comment by stephanbro** [comment]
@dagar Thanks for the info! I will definitely do that as I update this PR and fix the build issues.

@LorenzMeier I'm definitely not getting the full FW parameters, but I'm still getting a couple. I tried changing that specific parameter (FW_ARSP_MODE) to use the normal param find and it still shows up, so I don't think my changes is to blame for that.

One issue that remains is that some build configurations use the sdlog2, others the logger, and some both sdlog2 and the logger, but both modules use the SDLOG_UTC_OFFSET parameter. If the parameter is in the logger's parameters then it's not built when only building sdlog2, but if it's in both parameter's list there will be a duplicate.

How should this situation be handled? The generation script pruning the duplicate entries could be one options. Another option could be to have parameters that depend on build configuration use the standard `param_find`, but that parameter would work of course.

**Comment by jgoppert** [comment]
@stephanbro let me know if you have time to sync up on Skype or something.  I'm working on the hashtable param backend to support the ROS param API and want to make sure our work is compatible.

**Comment by stephanbro** [comment]
@jgoppert that sounds good. I sent you an email and we can talk about what makes the most sense.

**Comment by stephanbro** [comment]
I talked with @jgoppert and it looks like there won't be any issues with our work.

Is any one available to review? Maybe @dagar?

File: Tools/px_generate_params.py [hunk 1]
```diff
@@ -51,10 +54,14 @@
 				continue
 			params.append(param)
 
+i=0
```
Comment (dagar) [inline Tools/px_generate_params.py]: PEP8 spacing

**Comment by dagar** [comment]
I can take a look and test on nuttx.

 - So how are we going to know to use param_find() instead of PARAM_FIND()?
 - Would you mind squashing the revert commits?
 - Have you done any profiling? Are you familiar with how to run a SITL simulation? `make posix_sitl_default jmavsim`.
 - Big commits across the entire codebase can be painful for downstream consumers to merge. We should split the param_find -> PARAM_FIND changes into a commit per module. We can script it.

**Comment by bkueng** [comment]
@dagar raised some good points. I think we should strive for a single `param_find` method. a) the rules when to use which are too complicated and b) even if the rules were simple, there needs to be documentation (& method naming) which one to use.

Maybe we should disable the cmake scoping. The 'same param used in multiple modules' (logger, sdlog2, etc) issue is still unresolved, right? As our typical targets compile most modules, we don't gain much by the scoping anyway. In addition, https://github.com/PX4/Firmware/issues/6263 is still open.
That should also simplify the rules, when to use param_find vs PARAM_FIND.

Taking it one step further, we could do some compile-time magic to check whether it's a dynamic string or a literal. Something like [0] might work?

Another point is that PARAM_FIND will not call `param_set_used_internal`, which for instance is used to determine which parameters are sent via mavlink to a GCS (the point raised above by Lorenz). Is this resolved? If so, how?

[0] http://stackoverflow.com/questions/14925635/ways-to-distinguish-string-literal-and-runtime-generated-string

**Comment by davids5** [comment]
Hi,

I have lost context on this, so these comments may not be applicable. 
The original intent of the scoping was to fit code on a small FLASH footprint part that only had only 3 parameters but carried the burden of all 700+ preventing it from fitting in FLASH (px4cannode-v1). - This is still needed. 

The other importing thing that needs to be supported is the storages in SoC flash as on aereo. flash_params

**Comment by stephanbro** [comment]
@bkueng Disabling the cmake scoping would fix some issues like the one you mentioned with sdlog2 and the logger using the same parameter, but it's possible for only one to be built. It sounds like that would create some additional issues according to @davids5, so I'm not sure what the best solution is.

You make another good point about `param_set_used_internal` not being called and I noticed issues when I was connecting to QGC. My thoughts on the solution for that is to have QGC use the SYS_AUTOSTART to determine what parameters to show instead of a crc32 of what parameters are used, but I have limited insight into what makes the most sense for QGC.

Overall, I'm not confident that the improvement from these changes outweigh the potential issues that it creates. It may make the most sense to close this PR for now and revisit this at a later time.

**Comment by bkueng** [comment]
Seems like we need scoping then.

> One option could be to get more complicated with the "scope" and only generate the parameters that are actually used by parsing each file for a PARAM_FIND and making a list of those parameters

But you would miss the dynamic strings.

We could do this:
- First move to JSON/XML for param definition (https://github.com/PX4/Firmware/issues/6115), something like one JSON file per param group.
- let each module define a list on which param groups it depends on (that's more manual but more robust as well). That would replace the cmake scoping and resolve its current issues.
- then come back to this

What do you think?

**Comment by LorenzMeier** [comment]
I think there are way too many things mixed in here. What I see as dominant value is to replace the string handle at compile time with a shorter, unique representation. Since we don't have a static list we need to use something unique but dynamic, e.g. a 16 bit hash of the param name. We could even check that there are no hash collisions during the param XML generation.

Making the assumption about a static list of parameters in memory is not valid and not desired. But reducing all those strings to 2 bytes instead would have tremendous value in terms of runtime flash.

**Comment by jgoppert** [comment]
@LorenzMeier have you seen https://github.com/PX4/Firmware/pull/6305. I'm using HashMap which uses strings as the key and has a string hashing functiion. Similar to what you are suggesting but you don't keep the hash around, just the strings. We should be able to save a lot of memory this way.

**Comment by LorenzMeier** [comment]
I was mistaken. The string duplicate removal already realised pretty much most of the wins.

**Comment by stephanbro** [comment]
It looks as though it makes more sense to use something more dynamic like the HashMap in #6305 instead of this PR which is a lot more static, so I'm closing this for now. I can reopen it later if it's decided that this implementation makes sense.

**Pull request closed without merging**